state/table.c

```diff
@@ -0,0 +1,102 @@
+                                         uint64_t timeout,
+                                         void *data,
+                                         table_done_cb done_cb,
+                                         table_fail_cb fail_cb,
+                                         table_retry_cb retry_cb,
+                                         void *user_context) {
+  CHECK(db_handle);
+  CHECK(db_handle->loop);
+  /* Allocate and initialize callback data structure for object table */
+  table_callback_data *cb_data = malloc(sizeof(table_callback_data));
+  CHECKM(cb_data != NULL, "Memory allocation error!")
+  cb_data->id = id;
+  cb_data->done_cb = done_cb;
+  cb_data->fail_cb = fail_cb;
+  cb_data->retry_cb = retry_cb;
+  cb_data->retry_count = retry_count;
+  cb_data->timeout = timeout;
+  cb_data->data = data;
+  cb_data->requests_info = NULL;
+  cb_data->user_context = user_context;
+  cb_data->db_handle = db_handle;
+  /* Add timer and initialize it. */
+  cb_data->timer_id = event_loop_add_timer(db_handle->loop,
+                                           timeout,
+                                           table_timeout_handler,
+                                           cb_data);
+  outstanding_callbacks_add(cb_data);
+  return cb_data;
+}
+
+void destroy_table_callback(table_callback_data *cb_data) {
+  CHECK(cb_data != NULL);
+
+  if (cb_data->data)
+    free(cb_data->data);
+
+  if (cb_data->requests_info)
+    free(cb_data->requests_info);
+
+  outstanding_callbacks_remove(cb_data);
+
+  /* Timer is removed via EVENT_LOOP_TIMER_DONE in the timeout callback. */
+  free(cb_data);
+}
+
+int64_t table_timeout_handler(event_loop *loop,
+                              int64_t timer_id,
+                              void *user_context) {
+  CHECK(loop != NULL)
+  CHECK(user_context != NULL)
+  table_callback_data *cb_data = (table_callback_data *) user_context;
+
+  CHECK(cb_data->retry_count >= 0)
+  LOG_DEBUG("retrying operation, retry_count = %d", cb_data->retry_count);
+
+  if (cb_data->retry_count == 0) {
+    /* We didn't get a response from the database after exhausting all retries;
+     * let user know, cleanup the state, and remove the timer. */
+    cb_data->fail_cb(cb_data->id, cb_data->user_context);
+    destroy_table_callback(cb_data);
+    return EVENT_LOOP_TIMER_DONE;
+  }
+
+  /* Decrement retry count and try again. */
+  cb_data->retry_count--;
+  cb_data->retry_cb(cb_data);
+  return cb_data->timeout;
+}
+
+static outstanding_callback *outstanding_cbs = NULL;
+
+void outstanding_callbacks_add(table_callback_data *key) {
+  outstanding_callback *outstanding_cb = malloc(sizeof(outstanding_callback));
+
+  CHECK(outstanding_cb != NULL);
+  outstanding_cb->key = key;
+  HASH_ADD_PTR(outstanding_cbs, key, outstanding_cb);
+}
+
+outstanding_callback *outstanding_callbacks_find(table_callback_data *key) {
+  outstanding_callback *outstanding_cb = NULL;
+
+  HASH_FIND_PTR(outstanding_cbs, &key, outstanding_cb);
```

Comment:
What happens if we malloc something later on that gets allocated the same pointer?